src/NodeCompiler/CompileNodeToValue.php

```diff
@@ -47,6 +47,10 @@ public function __invoke(Node $node, CompilerContext $context)
             return $this->compileBinaryOperator($node, $context);
         }
 
+        if ($node instanceof Node\Scalar\MagicConst\Dir) {
+            return '__DIR__';
```

Comment:
If we are in the context of a class, `$context->getSelf()` returns the class, from which you can get the filename. Currently wouldn't work for a function though. It wouldn't be too difficult to also inject a filename though into `CompilerContext` constructor.